### PR TITLE
Add support for tolerations, env, annotations and labels

### PR DESCRIPTION
**Pull request by blublinsky**
This is a replacement (simplified) for ithub.com/ray-project/kuberay/pull/1066

## Why are these changes needed?
It is quite important to be able to specify environment variables, annotations, labels, and tolerations to control pod's behavior. This is exposed in Kuberay operator but not in the API server

The basic changes are:

Extending cluster and config protos to add new variables and regenerate go client
In the API server project modified podTemplateSpec building in utils cluster to add new values
In the API server project modified template config map creation in utils cluster to add new values
In the API server project modified CRD to API conversion in model converter to create new values

## Related issue number

Multiple issues related to adding additional parameters support to API server

## Checks

- [ ] I've made sure the tests are passing. 
- Testing Strategy
   - [ ] Unit tests
   - [x ] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by blublinsky** [comment]
Hi @kevin85421,
replaced the previous PR with a new one. Reverted to go 1.17

**Comment by blublinsky** [comment]
Hi @kevin85421,
Any progress on accepting this one? It looks like its passing all checks and I am using it internally for my work without any issues

**Comment by kevin85421** [comment]
Hi @blublinsky, thank you for the contribution! KubeRay APIServer is a totally community-managed component, so I will let @Jeffwan @scarlet25151 review this PR. In addition, as I mentioned https://github.com/ray-project/kuberay/pull/1066#pullrequestreview-1415179565, it will be great to have tests to make sure the functionality.

File: apiserver/pkg/util/cluster.go
Comment (rjmco) [inline apiserver/pkg/util/cluster.go:48]: Typo

Comment (blublinsky) [reply apiserver/pkg/util/cluster.go:48]: Fixed

Comment (blublinsky) [reply apiserver/pkg/util/cluster.go:48]: Fixed

**Review by z103cb — COMMENTED** [review]
@blublinsky you should probably cover the newly added coded with unit-tests, focusing primarely on error handling.

File: apiserver/pkg/model/converter.go
Comment (blublinsky) [reply apiserver/pkg/model/converter.go]: fixed

**Comment by blublinsky** [comment]
Now with unit tests

File: proto/cluster.proto
Comment (scarlet25151) [inline proto/cluster.proto:215]: @blublinsky in the L166 there is a envs for the whole cluster, so here I would like to double comfirm that is this used for specific config of head and worker, in this situation head and worker may have some different envs/label/annotation?

**Review by scarlet25151 — APPROVED** [review]
overall /LGTM.

File: proto/cluster.proto
Comment (blublinsky) [reply proto/cluster.proto:215]: Yes, they are specified independently for head and different worker groups

**Comment by blublinsky** [comment]
@kevin85421 now that it is approved, can we, please, merge it

**Comment by kevin85421** [comment]
@blublinsky Cool, would you mind fixing the lint error in CI? Thanks!

**Comment by blublinsky** [comment]
Hi @kevin85421 ,
The linter error is in the generated code:
```
Running [/home/runner/golangci-lint-1.29.0-linux-amd64/golangci-lint run --out-format=github-actions --path-prefix=./apiserver --timeout=3m] in [/home/runner/work/kuberay/kuberay/apiserver] ...
  Error: copylocks: call of t.Errorf copies lock value: github.com/ray-project/kuberay/proto/go_client.PodToleration contains google.golang.org/protobuf/internal/impl.MessageState contains sync.Mutex (govet)
  
  Error: issues found
  Ran golangci-lint in [35](https://github.com/ray-project/kuberay/actions/runs/5011590782/jobs/9058495484?pr=1070#step:8:37)614ms
```
It probably should be just excluded from the linter. Any suggestions on how to do this?

File: apiserver/pkg/model/converter_test.go
Comment (z103cb) [inline apiserver/pkg/model/converter_test.go:238]: use the %d, for integer.

File: apiserver/pkg/util/cluster_test.go
Comment (blublinsky) [reply apiserver/pkg/util/cluster_test.go:228]: fixed

**Comment by kevin85421** [comment]
Hi @blublinsky, could you fix the lint error? Thanks!

**Comment by blublinsky** [comment]
@kevin85421 I think I got them all now. I had to change a lot of files because golangci-lint run wants `klog` to be explicitly defined like `klog "k8s.io/klog/v2"`

**Comment by blublinsky** [comment]
The failures here are in serve and job tests, not my code

**Comment by kevin85421** [comment]
cc @scarlet25151, there have been several commits after your approval of this PR. Would you mind reviewing it again before I merge it? Thanks!

**Comment by blublinsky** [comment]
@kevin85421 can we, please, finally merge it?
I want to do several more PRs, but am trying to do one at a time

**Comment by kevin85421** [comment]
Hi @blublinsky, thank you for the follow-up. The KubeRay API Server is fully managed by the community, not by Anyscale. As I am not familiar with this component, I am unable to review it. I will merge this PR only when the owners approve it. Thanks!

**Comment by kevin85421** [comment]
Gentle ping: @scarlet25151 @Jeffwan would you mind reviewing this PR? Thanks!

File: apiserver/pkg/model/converter.go
Comment (scarlet25151) [inline apiserver/pkg/model/converter.go:18]: nit: In the future, we can list the constant string below separately, like:https://github.com/ray-project/kuberay/blob/25864689ba31214bfe3469b762d2dac29ce72f89/apiserver/pkg/util/config.go#L11-L18

File: apiserver/pkg/model/converter.go
Comment (scarlet25151) [inline apiserver/pkg/model/converter.go:42]: @blublinsky would you please help on list them separately in a style of below, and give a comment of this ENVs usage in this or an additional PR?
```go
const (
	// used to indicate if we need to disable docker cpu warning
	RAY_DISABLE_DOCKER_CPU_WARNING = "RAY_DISABLE_DOCKER_CPU_WARNING"
)
```

File: apiserver/pkg/model/converter.go
Comment (scarlet25151) [inline apiserver/pkg/model/converter.go:124]: I've seen some overlap of the default annotation and labels between node and ray cluster, here the reason to delete them is?

File: apiserver/pkg/util/cluster.go
Comment (scarlet25151) [inline apiserver/pkg/util/cluster.go:183]:
```suggestion
	// Add GPU resource support and append specific environment parameters for ray head.
	// TODO(@blublinsky): need smarter algorithm to filter main container. for example filter by name `ray-worker`
```

File: apiserver/pkg/util/cluster.go
Comment (scarlet25151) [inline apiserver/pkg/util/cluster.go:383]: Ditto
```suggestion
	// Add GPU resource support and append specific environment parameters for ray worker.
	if container, index, ok := GetContainerByName(podTemplateSpec.Spec.Containers, "ray-worker"); ok {
```

**Review by scarlet25151 — APPROVED** [review]
I left few comment, @blublinsky PTAL, and for some constant definition would you please change the format to align with https://github.com/ray-project/kuberay/blob/master/apiserver/pkg/util/config.go in the following PRs?

File: apiserver/pkg/model/converter.go
Comment (blublinsky) [reply apiserver/pkg/model/converter.go:18]: Done

File: apiserver/pkg/util/cluster.go
Comment (blublinsky) [reply apiserver/pkg/util/cluster.go:183]: Done

File: apiserver/pkg/util/cluster.go
Comment (blublinsky) [reply apiserver/pkg/util/cluster.go:383]: Done

**Comment by kevin85421** [comment]
@blublinsky feel free to ping me when you address all the comments. Thanks!

**Comment by blublinsky** [comment]
@kevin85421, @scarlet25151  Comments addressed

**Review by scarlet25151 — APPROVED** [review]
Overall LGTM.

**Comment by kevin85421** [comment]
Thank @blublinsky for the contribution and @scarlet25151 for the review! I will merge this PR after all CI tests pass.

**Comment by kevin85421** [comment]
@blublinsky I knew it, but we still need to pass the flaky tests before merging every PR.

**Review by kevin85421 — APPROVED** [review]
Merge this PR as it has received approval from two reviewers. I have not personally reviewed it.

**Pull request merged**
